### PR TITLE
Manage members for an integration (BE)

### PR DESCRIPTION
**Pull request by JedrzejJanasiak**
## Issue

https://gravitee.atlassian.net/browse/APIM-6135

## Description

PR provides new endpoints to manage members of an integration. 
Additionally, a list of integration and integration details now returns also information about the Primary Owner. 

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-zkhqggepve.chromatic.com)

### CODE REVIEW

File: gravitee-apim-rest-api/gravitee-apim-rest-api-rest/src/main/java/io/gravitee/rest/api/rest/filter/PermissionsFilter.java [hunk 1]
```diff
@@ -91,6 +92,8 @@ private boolean hasPermission(Permission permission, ContainerRequestContext req
                 return hasPermission(executionContext, permission, getApiId(requestContext));
             case GROUP:
                 return hasPermission(executionContext, permission, getGroupId(requestContext));
+            case INTEGRATION:
+                return hasPermission(executionContext, permission, getId(INTEGRATION_ID_PARAM, requestContext));
```
Comment (michel-barret) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-rest/src/main/java/io/gravitee/rest/api/rest/filter/PermissionsFilter.java]: We can use switch expression

```java
return switch() {
  case ORGANISATION -> hasPermission(executionContext, permission, executionContext.getOrganizationId());
  // ...
};
```

and replace default by the last one none managed `PLATEFORM`

Comment (JedrzejJanasiak) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-rest/src/main/java/io/gravitee/rest/api/rest/filter/PermissionsFilter.java]: 👍

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/integration/IntegrationResource.java [hunk 2]
```diff
@@ -246,4 +252,9 @@ public DeletedIngestedApisResponse deleteIngestedApis(@PathParam("integrationId"
 
         return new DeletedIngestedApisResponse().deleted(output.deleted()).skipped(output.skipped()).errors(output.errors());
     }
+
+    @Path("/members")
+    public IntegrationMembersResource getApiMembersResource() {
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/integration/IntegrationResource.java]:
```suggestion
    public IntegrationMembersResource getIntegrationMembersResource() {
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/membership/domain_service/IntegrationPrimaryOwnerDomainService.java [hunk 3]
```diff
@@ -34,8 +39,70 @@
...
 
+    public PrimaryOwnerEntity getIntegrationPrimaryOwner(final String organizationId, String integrationId)
+        throws IntegrationPrimaryOwnerNotFoundException {
+        return findPrimaryOwnerRole(organizationId)
+            .flatMap(role ->
+                findIntegrationPrimaryOwnerMembership(integrationId, role)
+                    .flatMap(membership ->
+                        switch (membership.getMemberType()) {
+                            case USER -> findUserPrimaryOwner(membership);
+                            case GROUP -> findGroupPrimaryOwner(membership, role.getId());
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/membership/domain_service/IntegrationPrimaryOwnerDomainService.java]: With the new decision we can't have group primary owner. I suppose we can simplify that

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/membership/exception/IntegrationPrimaryOwnerNotFoundException.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.apim.core.membership.exception;
+
+import lombok.Getter;
+
+@Getter
+public class IntegrationPrimaryOwnerNotFoundException extends RuntimeException {
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/core/membership/exception/IntegrationPrimaryOwnerNotFoundException.java]:
```suggestion
public class IntegrationPrimaryOwnerNotFoundException extends NotFoundDomainException {
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/integration/use_case/DeleteIntegrationUseCaseTest.java [hunk 5]
```diff
@@ -37,14 +46,24 @@ public class DeleteIntegrationUseCaseTest {
 
     IntegrationCrudServiceInMemory integrationCrudServiceInMemory = new IntegrationCrudServiceInMemory();
     ApiQueryServiceInMemory apiQueryServiceInMemory = new ApiQueryServiceInMemory();
+    MembershipCrudServiceInMemory membershipCrudService = new MembershipCrudServiceInMemory();
+    MembershipQueryServiceInMemory membershipQueryService = new MembershipQueryServiceInMemory(membershipCrudService);
+    AuditCrudServiceInMemory auditCrudService = new AuditCrudServiceInMemory();
+    UserCrudServiceInMemory userCrudService = new UserCrudServiceInMemory();
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/integration/use_case/DeleteIntegrationUseCaseTest.java:54]: we should reset them in `afterEach`

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/membership/domain_service/IntegrationPrimaryOwnerDomainServiceTest.java [hunk 6]
```diff
@@ -50,6 +60,9 @@ public class IntegrationPrimaryOwnerDomainServiceTest {
     private static final AuditInfo AUDIT_INFO = AuditInfoFixtures.anAuditInfo(ORGANIZATION_ID, ENVIRONMENT_ID, USER_ID);
 
     MembershipCrudServiceInMemory membershipCrudService = new MembershipCrudServiceInMemory();
+    MembershipQueryServiceInMemory membershipQueryService = new MembershipQueryServiceInMemory(membershipCrudService);
+    UserCrudServiceInMemory userCrudService = new UserCrudServiceInMemory();
+    GroupQueryServiceInMemory groupQueryService = new GroupQueryServiceInMemory();
```
Comment (jgiovaresco) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/test/java/io/gravitee/apim/core/membership/domain_service/IntegrationPrimaryOwnerDomainServiceTest.java]: we should reset them in afterEach

**Review by jgiovaresco — APPROVED** [review]

**Pull request merged**